sdccc/src/main/java/com/draeger/medical/sdccc/sdcri/testclient/TestClientImpl.java

```diff
@@ -354,17 +405,90 @@ void deviceFound(final ProbedDeviceFoundMessage message) {
         }
 
         sdcRemoteDevice.registerWatchdogObserver(this);
+        isConnected.set(true);
+    }
+
+    @Override
+    public synchronized CompletableFuture<Boolean> enableReconnect(final long timeoutInSeconds) {
+        final var actualTimeout = getReconnectFeatureTimeout(timeoutInSeconds);
+        LOG.debug("Reconnect enabled for {} seconds.", actualTimeout);
+
+        this.reconnectEnabled.set(true);
+        final var ctx = (LoggerContext) LogManager.getContext(this.getClass().getClassLoader(), false);
+        final var appender =
+                (TriggerOnErrorOrWorseLogAppender) ctx.getConfiguration().getAppender(APPENDER_NAME);
+        appender.setThreadNameWhitelist(buildThreadNameWhiteList());
+
+        this.reconnectFuture = new CompletableFuture<>();
+
+        reconnectFuture.whenComplete((result, exception) -> {
+            if (exception != null) {
+                LOG.error("Something went wrong during reconnect.", exception);
+                testRunObserver.invalidateTestRun(exception.getMessage());
+            } else if (!isConnected.get()) {
+                LOG.debug("No connection established during reconnect process, invalidating the test run.");
+                testRunObserver.invalidateTestRun(COULDN_T_RECONNECT);
+            }
+            appender.setThreadNameWhitelist(List.of());
+            this.inReconnectProcess.set(false);
+            this.reconnectEnabled.set(false);
+        });
+
+        final ScheduledExecutorService timeoutScheduler = Executors.newSingleThreadScheduledExecutor();
+        try {
+            timeoutScheduler.schedule(
+                    () -> {
```

Comment:
This will potentially run after subsequent enableReconnect calls, especially when waiting for the future as intended and getting a value out of it due to a reconnect happening.
When that happens the reconnectEnabled state will be broken when the subsequent enableReconnect reset reconnectEnabled, but no completion related to the second enableReconnect got triggered. The first scheduled task shares the instance of reconnectFuture with the second one temporarily in this case, which is a design flaw.
I even wrote a test case trying to use the new features in the intended way twice in a sequential manner and it ran into exactly this issue from which it could not recover.